docs/guides/javascript.rst

```diff
@@ -12,88 +12,159 @@ The 1.8 version is still functional and is :ref:`described below<1.8-js>`.
+
+Below we refactor a bit so that the module depends on a new ``myplugin/hello`` module to provide
+the greeting:
 
 .. code-block:: javascript
 
-	define(function(require) {
-		var elgg = require("elgg");
-		var $ = require("jquery");
+    // in views/default/js/myplugin/hello.js
 
-		return function() {
-			// Some logic in here
-		};
-	});
+    define(function(require) {
+        var elgg = require("elgg");
 
-Define your module via a URL
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+        return elgg.echo('hello_world');
+    });
 
-You can define an existing AMD module using ``elgg_define_js()``. Traditional (browser-globals)
-JavaScript files can also be defined as AMD modules if you shim them by setting ``exports`` and
-optionally ``deps``.
+.. code-block:: javascript
+
+    // in views/default/js/myplugin/say_hello.js
+
+    define(function(require) {
+        var $ = require("jquery");
+        var hello = require("myplugin/hello");
+
+        $('body').append(hello);
+    });
 
-.. warning:: Calls to ``elgg_define_js()`` must be in an ``init, system`` event handler.
+
+Defining a module with a PHP view
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Although plain JavaScript files are recommended, you may register a PHP view as an AMD module.
+
+To create the same module as above, you'd create the file ``views/default/js/myplugin/say_hello.js.php``
+(note the double extension ``.js.php``).
+
+You must also manually register the view as an external resource:
 
 .. code-block:: php
 
     <?php
+    // note the view name includes the .js extension, but not .php
+    elgg_register_external_view('js/myplugin/say_hello.js', true);
 
-	elgg_register_event_handler('init', 'system', 'amd_init');
+.. note::
 
-	function amd_init() {
-		// AMD module as the view js/backbone
-		elgg_define_js('backbone');
+    Although the view is PHP, it's still cached, so you should treat the output as static (the same for
+    all users).
 
-		// AMD module with a different path
-		elgg_define_js('backbone', array(
-			'src' => '/vendors/backbone/backbone.js',
-		));
+Setting the URL of a module
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-		// Shimmed AMD module
-		elgg_define_js('jquery.form', array(
-			'src' => '/vendors/jquery/jquery.form.js',
-			'deps' => array('jquery'),
-			'exports' => 'jQuery.fn.ajaxForm',
-		));
-	}
+You may have a script outside your views you wish to make available as a module.
 
-Some things to note
-^^^^^^^^^^^^^^^^^^^
+In your PHP ``init, system`` event handler, you can use ``elgg_define_js()`` to do this:
 
-1. Do not use ``elgg.provide()`` or ``elgg.require()`` anymore. They are fully replaced by ``define()`` and ``require()`` respectively.
-2. Return the value of the module instead of adding to a global variable.
-3. Static views (.css, .js) are automatically minified and cached by Elgg's simplecache system.
+.. code-block:: php
+
+    <?php
+    elgg_define_js('myplugin/somelib', [
+        'src' => '/mod/myplugin/vendors/somelib.js',
```

Comment:
Specifically, pointing directly at files in the document root seems bad. We do it pervasively now and I don't have a good  solution, but it seems like it is locking us in to a very particular structure that we might not want...